### PR TITLE
Implement DevFlags for DSC API

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This PR adds implementation to support `custom manifests` for components, that can be pulled directly from component repos.

## How Has This Been Tested?
- Dev Build: **quay.io/vhire/opendatahub-operator-catalog:v2.3.0-1**
- Custom Manifests Example
```
    dashboard:
      managementState: Managed
      devFlags:
        manifests:
          - manifestsFolder: manifests
            uri: >-
              https://github.com/opendatahub-io/odh-dashboard/tarball/v2.15.0-incubation
```
- Custom Manifests with Overlays
```
    dashboard:
     managementState: Managed
      devFlags:
        manifests:
          - manifestsFolder: manifests
            overlay: overlays/incubation
            uri: >-
              https://github.com/opendatahub-io/odh-dashboard/tarball/v2.15.0-incubation
```
- Example of Component with references to multiple repos
```
    kserve:
      managementState: Managed
      devFlags:
        manifests:
          - manifestsFolder: config
            overlay: overlays/odh
            uri: 'https://github.com/opendatahub-io/kserve/tarball/master'
          - manifestsFolder: config
            uri: >-
              https://github.com/opendatahub-io/odh-model-controller/tarball/main
```

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by zdtsw** [comment]
i have a strange feeling, did this PR miss to commit the change of API file?  `datasciencecluster_types.go`

File: components/codeflare/codeflare.go
Comment (zdtsw) [inline components/codeflare/codeflare.go:21]:
```suggestion
	CodeflarePath       = deploy.DefaultManifestPath + "/"+ ComponentName + "/base"
```

File: components/modelmeshserving/modelmeshserving.go
Comment (zdtsw) [inline components/modelmeshserving/modelmeshserving.go:36]: maybe we can leave `model-mesh` out of the `devFlags` for now, till they are done with transition, then we know how they want to use this: it could be one repo , but most likely will be two different one for modelmesh one for monitoring

File: components/ray/ray.go
Comment (zdtsw) [inline components/ray/ray.go:16]:
```suggestion
	RayPath       = deploy.DefaultManifestPath + "/" + ComponentName + "/operator/base"
```

File: components/codeflare/codeflare.go
Comment (zdtsw) [inline components/codeflare/codeflare.go:32]:
```suggestion
func (c *CodeFlare) DownloadManifests(_ string) error {
```

File: components/datasciencepipelines/datasciencepipelines.go
Comment (zdtsw) [inline components/datasciencepipelines/datasciencepipelines.go:32]:
```suggestion
func (d *DataSciencePipelines) DownloadManifests(_ string) error {
```

File: components/modelmeshserving/modelmeshserving.go
Comment (zdtsw) [inline components/modelmeshserving/modelmeshserving.go:34]:
```suggestion
func (m *ModelMeshServing) DownloadManifests(_ string) error {
```

File: components/ray/ray.go
Comment (zdtsw) [inline components/ray/ray.go:27]:
```suggestion
func (r *Ray) DownloadManifests(_ string) error {
```

File: components/workbenches/workbenches.go
Comment (zdtsw) [inline components/workbenches/workbenches.go:38]: 2 parts need to consider:
- after the transition, workbench will only use "kubeflow" git repo,
- user who forked the kubeflow can have a different repo name, e.g
```
workbenches:
      devFlags:
        manifests:
          - manifestsFolder: component/notebook-controller
            overlay: base
            uri: 'https://github.com/my-repo/my-kubeflow/tarball/some-branch'
          - manifestsFolder: component/odh-notebook-controller
            overlay: overlays/openshift
            uri: >-
              https://github.com/your-repo/your-kubeflow/tarball/another-branch
```

any reason we need to  have this check on `strings.Contains(subcomponent.URI,...)`

Comment (zdtsw) [reply components/workbenches/workbenches.go:38]: in my PR to get nbc working after trantiion,
i renamed it from "notebook-controller" to the folder "kf-notebook-controller" in the operator image.
e.g 
`    "kubeflow:v1.7-branch:components/notebook-controller/config:odh-notebook-controller/kf-notebook-controller"`

so for workbench, the case might be to check s`trings.Contains(manifestsFolder, "odh-notebook-controller")`
and `strings.Contains(manifestsFolder, "kf-notebook-controller")`

File: controllers/dscinitialization/dscinitialization_controller.go
Comment (zdtsw) [inline controllers/dscinitialization/dscinitialization_controller.go:120]: do we want to remove this function totally? 
- modelmesh has not done transition, if we remove it, we can only merge this PR after they  finish the transion or even till we announce odh-manifests repo is archived 
- need put some word in the DSCI API saying this is not in use any more.

File: components/component.go
Comment (VaishnaviHire) [reply components/component.go:60]: The platform is needed by some components to update the [path](https://github.com/opendatahub-io/opendatahub-operator/pull/538/files/a1d3b061acbaee1d6636a9f32d280fed700aab0e#diff-3ddcc8f65a628050fbf81b0ede482ebe52b1418631bf8586c509916b8b767dcdR52)

File: components/dashboard/dashboard.go
Comment (VaishnaviHire) [reply components/dashboard/dashboard.go:46]: Yes, did this to ensure multiple manifests are not added for a component that doesn't require them.

File: components/workbenches/workbenches.go
Comment (VaishnaviHire) [reply components/workbenches/workbenches.go:38]: This implementation doesn't support multiple versions of same manifests, we have tight coupling with deployment paths.

The manifests list just state if there are multiple sub-components, example `kubeflow` and `notebooks`, a user can list them.

**Review by zdtsw — APPROVED** [review]
did test on this catalogsoure image
something is not doing as expected.

copied with example from dashboard
Screenshot from 2023-09-21 08-54-09
saved it and come back again to check
Screenshot from 2023-09-21 08-54-40


similar if i do the change from "Form view" which is easy to just fill in all the info, no need to check "spec",
create it then check again, it is gone.

File: components/component.go
Comment (zdtsw) [inline components/component.go:35]:
```suggestion
	// List of custom manifests for the given component
```

File: components/component.go
Comment (zdtsw) [inline components/component.go:41]:
```suggestion
	// uri is the URI for git repo.
```

File: components/component.go
Comment (zdtsw) [inline components/component.go:45]:
```suggestion
	// manifestsFolder is the relative path to the folder containing manifests in a repository.
```

File: components/component.go
Comment (zdtsw) [inline components/component.go:49]:
```suggestion
	// overlay is the subpath within manifestsFolder where kustomize builds start.
```

**Comment by VaishnaviHire** [comment]
@zdtsw Updated to include kserve(pull from odh-manifests) and workbench(2 separate paths for controllers) changes

File: get_all_manifests.sh
Comment (zdtsw) [inline get_all_manifests.sh:32]: do we need this for workbench? see line 13

Comment (zdtsw) [reply get_all_manifests.sh:32]: think you will need a rebase from "main" to get the change, like line 11 and 14 are duplicated.

**Comment by VaishnaviHire** [comment]
@zdtsw Final update with changes in devflag spec

Neew spec-
```
       devFlags:
          manifests:
             - uri: <github.repo>/tarball/<branch/tag>
               contextDir: <path to manifest folder>
               sourcePath: <path to start kustomize build> -----------------> This can be `base`, `overlays/x`, `default`, `custom` etc
```

**Comment by VaishnaviHire** [comment]
Merging this, since it has approval from @zdtsw . Latest changes including updates in naming

**Pull request merged**
